### ISSUE TITLE
Inconsistent data loading (spacy debug data)

### ISSUE THREAD

**Issue by kormilitzin** [explosion/spaCy#8266]
## How to reproduce the behaviour

I am experimenting with custom NER training. I generated both training and dev data sets, however when I run `spacy training` and `debug data` commands, the output shows only a small fraction of training instances and all of the dev set.

Specifically, using **100** training example and **50** dev examples and running the following command:

`python -m spacy debug data ./ner_config/config.cfg --paths.train ./data/spacy_format/train.spacy --paths.dev ./data/spacy_format/dev.spacy
`

The output:

> ============================ Data file validation ============================
> ✔ Corpus is loadable
> ✔ Pipeline can be initialized with data
> 
> =============================== Training stats ===============================
> Language: en
> Training pipeline: transformer, ner
> 3 training docs
> 50 evaluation docs
> ⚠ 2 training examples also in evaluation data
> ✘ Low number of examples to train a new pipeline (3)
> 
> ============================== Vocab & Vectors ==============================
> ℹ 893 total word(s) in the data (218 unique)
> ℹ No word vectors present in the package
> 
> ========================== Named Entity Recognition ==========================
> ℹ 1 label(s)
> 0 missing value(s) (tokens with '-' label)
> ⚠ Low number of examples for label 'DRUG' (19)
> huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
> To disable this warning, you can either:
>         - Avoid using `tokenizers` before the fork if possible
>         - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
> ⚠ No examples for texts WITHOUT new label 'DRUG'
> ✔ No entities consisting of or starting/ending with whitespace
> ✔ No entities consisting of or starting/ending with punctuation
> 
> ================================== Summary ==================================
> ✔ 4 checks passed
> ⚠ 3 warnings
> ✘ 1 error

When I load 100 training instances (through DocBin()) saved to a disc to python IDE, I can get all 100 examples and their annotations (`doc.ents`)

When I used **the same data set** for both `trains` and dev`, I expected to get **the same** number of training and dev examples:

`python -m spacy debug data ./ner_config/config.cfg --paths.train ./data/spacy_format/**dev.spacy** --paths.dev ./data/spacy_format/**dev.spacy**`

however, I got the same pattern of only a small fraction of training examples (2) and all examples in the dev set (50):

> =========================== Data file validation ===============================
> ✔ Corpus is loadable
> ✔ Pipeline can be initialized with data
> 
> =============================== Training stats ===============================
> Language: en
> Training pipeline: transformer, ner
> **2 training docs**
> **50 evaluation docs**
> ⚠ 2 training examples also in evaluation data
> ✘ Low number of examples to train a new pipeline (2)
> 
> ============================== Vocab & Vectors ==============================
> ℹ 696 total word(s) in the data (188 unique)
> ℹ No word vectors present in the package
> 
> ========================== Named Entity Recognition ==========================
> ℹ 1 label(s)
> 0 missing value(s) (tokens with '-' label)
> ⚠ Low number of examples for label 'DRUG' (17)
> huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
> To disable this warning, you can either:
>         - Avoid using `tokenizers` before the fork if possible
>         - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
> ⚠ No examples for texts WITHOUT new label 'DRUG'
> ✔ No entities consisting of or starting/ending with whitespace
> ✔ No entities consisting of or starting/ending with punctuation
> 
> ================================== Summary ==================================
> ✔ 4 checks passed
> ⚠ 3 warnings
> ✘ 1 error
> 

**What am I doing wrong here? How to troubleshoot it?
Thanks!**

## Info about spaCy

- **spaCy version:** 3.0.6
- **Platform:** Linux-5.4.0-72-generic-x86_64-with-glibc2.27
- **Python version:** 3.9.5
- **Pipelines:** en_core_web_trf (3.0.0), en_core_web_sm (3.0.0), en_core_web_lg (3.0.0)

## config.cfg file

[paths]
train = null
dev = null
vectors = null
init_tok2vec = null

[system]
gpu_allocator = "pytorch"
seed = 0

[nlp]
lang = "en"
pipeline = ["transformer","ner"]
batch_size = 512
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.ner]
factory = "ner"
moves = null
update_with_oracle_cut_size = 100

[components.ner.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "ner"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = false
nO = null

[components.ner.model.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0
pooling = {"@layers":"reduce_mean.v1"}
upstream = "*"

[components.transformer]
factory = "transformer"
max_batch_items = 4096
set_extra_annotations = {"@annotation_setters":"spacy-transformers.null_annotation_setter.v1"}

[components.transformer.model]
@architectures = "spacy-transformers.TransformerModel.v1"
name = "roberta-base"

[components.transformer.model.get_spans]
@span_getters = "spacy-transformers.strided_spans.v1"
window = 128
stride = 96

[components.transformer.model.tokenizer_config]
use_fast = true

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 500
gold_preproc = false
limit = 0
augmenter = null

[training]
accumulate_gradient = 3
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_padded.v1"
discard_oversize = true
size = 2000
buffer = 256
get_length = null

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001

[training.optimizer.learn_rate]
@schedules = "warmup_linear.v1"
warmup_steps = 250
total_steps = 20000
initial_rate = 0.00005

[training.score_weights]
ents_per_type = null
ents_f = 1.0
ents_p = 0.0
ents_r = 0.0

[pretraining]

[initialize]
vectors = null
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]

**Comment by polm** [comment]
That's really weird, thanks for the report. 

I assume your data is private, but if it isn't it'd be great if you could share it.

Either way we'll be looking into this.

**Comment by adrianeboyd** [comment]
This is most likely due to `max_length` in the training corpus. Try setting `max_length = 0`.

**Comment by kormilitzin** [comment]
@adrianeboyd and @polm -- thank you both. It was indeed the `max_length` issue. Now all sorted. 

Just out of curiosity - what was the reason to treat training and dev data sets differently?

**Comment by adrianeboyd** [comment]
We've changed the default to `max_length = 0` for upcoming versions because this is confusing. It's a good default for our internal training, but not for the default configs for users.

If the training data has sentence boundaries (which almost all of our internal training data does), the corpus reader splits up longer texts into sentences instead of skipping them, which makes for more even batches and fewer OOM issues around long documents while training.

For the dev evaluation, it's more important to keep the documents exactly in their original format, and the memory issues are less of a concern than while training.
